mlflow/langchain/retriever_chain.py

```diff
@@ -0,0 +1,139 @@
+"""Chain for wrapping a retriever."""
+from __future__ import annotations
+
+import json
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import yaml
+from langchain.callbacks.manager import AsyncCallbackManagerForChainRun, CallbackManagerForChainRun
+from langchain.chains.base import Chain
+from langchain.schema import BaseRetriever, Document
+from pydantic import Extra, Field
+from mlflow.utils.annotations import experimental
+
+
+@experimental
+class RetrieverChain(Chain):
+
+    """
+    Chain that wraps a retriever for use with MLflow.
+
+    The MLflow ``langchain`` flavor provides the functionality to log a retriever object and
+    evaluate it individually. This is useful if you want to evaluate the quality of the
+    relevant documents returned by a retriever object without directing these documents
+    through a large language model (LLM) to yield a summarized response.
+
+    In order to log the retriever object in the ``langchain`` flavor, the retriever object
+    needs to be wrapped within a ``RetrieverChain``.
+
+    See :ref:`log-retriever-chain` for how to log the ``RetrieverChain``.
+
+    :param retriever: The retriever to wrap.
+    """
+
+    input_key: str = "query"
+    output_key: str = "source_documents"
+    retriever: BaseRetriever = Field(exclude=True)
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+        arbitrary_types_allowed = True
+
+    @property
+    def input_keys(self) -> List[str]:
+        """Return the input keys."""
+        return [self.input_key]
+
+    @property
+    def output_keys(self) -> List[str]:
+        """Return the output keys."""
+        return [self.output_key]
+
```

Comment:
In fact, the main value of this `RetrieverChain` is the definition of the `input_keys` and `output_keys`.